common/task/weighted_channel_pool.go

```diff
@@ -0,0 +1,225 @@
+		//  0 -> 5
+		//  1 -> 3
+		//  2 -> 2
+		//  3 -> 1
+		// then iwrrChannels will contain chan [0, 0, 0, 1, 0, 1, 2, 0, 1, 2, 3] (ID-ed by channel key)
+		// This implementation uses interleaved weighted round robin schedule instead of classic weighted round robin schedule
+		// ref: https://en.wikipedia.org/wiki/Weighted_round_robin#Interleaved_WRR
+		iwrrSchedule atomic.Value // []*weightedChannel[V]
+	}
+)
+
+func NewWeightedRoundRobinChannelPool[K comparable, V any](timeSource clock.TimeSource, options WeightedRoundRobinChannelPoolOptions) *WeightedRoundRobinChannelPool[K, V] {
+	return &WeightedRoundRobinChannelPool[K, V]{
+		bufferSize:              options.BufferSize,
+		idleChannelTTLInSeconds: options.IdleChannelTTLInSeconds,
+		timeSource:              timeSource,
+		channelMap:              make(map[K]*weightedChannel[V]),
+		shutdownCh:              make(chan struct{}),
+	}
+}
+
+func (p *WeightedRoundRobinChannelPool[K, V]) Start() {
+	if !atomic.CompareAndSwapInt32(&p.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	p.shutdownWG.Add(1)
+	go p.cleanupLoop()
+}
+
+func (p *WeightedRoundRobinChannelPool[K, V]) Stop() {
+	if !atomic.CompareAndSwapInt32(&p.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	close(p.shutdownCh)
+	p.shutdownWG.Wait()
+}
+
+func (p *WeightedRoundRobinChannelPool[K, V]) cleanupLoop() {
+	defer p.shutdownWG.Done()
+	ticker := p.timeSource.NewTicker(time.Duration((p.idleChannelTTLInSeconds / 2)) * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.Chan():
+			p.doCleanup()
+		case <-p.shutdownCh:
+			return
+		}
+	}
+}
+
+func (p *WeightedRoundRobinChannelPool[K, V]) doCleanup() {
+	p.Lock()
+	defer p.Unlock()
+	var channelsToCleanup []K
+	now := p.timeSource.Now().Unix()
+	for k, v := range p.channelMap {
+		if now-v.lastAccessTime.Load() > p.idleChannelTTLInSeconds && len(v.c) == 0 && v.refCount.Load() == 0 {
+			channelsToCleanup = append(channelsToCleanup, k)
+		}
+	}
+
+	for _, k := range channelsToCleanup {
+		delete(p.channelMap, k)
+	}
+
+	if len(channelsToCleanup) > 0 {
+		p.updateScheduleLocked()
+	}
+}
+
+func (p *WeightedRoundRobinChannelPool[K, V]) GetOrCreateChannel(key K, weight int) (chan V, func()) {
+	p.RLock()
+	if v := p.channelMap[key]; v != nil && v.weight == weight {
+		v.refCount.Add(1)
+		v.lastAccessTime.Store(p.timeSource.Now().Unix())
+		p.RUnlock()
+		return v.c, func() {
+			v.refCount.Add(-1)
+		}
+	}
+	p.RUnlock()
+
+	p.Lock()
+	defer p.Unlock()
+	if v := p.channelMap[key]; v != nil {
+		v.refCount.Add(1)
+		v.lastAccessTime.Store(p.timeSource.Now().Unix())
+		if v.weight != weight {
+			v.weight = weight
+			p.updateScheduleLocked()
+		}
+		return v.c, func() {
+			v.refCount.Add(-1)
+		}
+	}
+
+	v := &weightedChannel[V]{
+		weight: weight,
+		c:      make(chan V, p.bufferSize),
+	}
+	p.channelMap[key] = v
+	v.refCount.Add(1)
+	v.lastAccessTime.Store(p.timeSource.Now().Unix())
```

Comment:
nit: rename this to `lastWriteTime` because `GetOrCreateChannel()` is only used to add a task to a channel. Reading is done via `GetSchedule()` or `GetAllChannels()`, and they don't update this timestamp.